minoc/src/main/java/org/opencadc/minoc/GetAction.java

```diff
@@ -134,15 +143,41 @@ public void doAction() throws Exception {
         storageLocation.storageBucket = artifact.storageLocation.storageBucket;
         
         log.debug("retrieving artifact from storage: " + storageLocation);
+
+        String range = syncInput.getHeader("Range");
+        log.debug("Range: " + range);
         
         ByteCountOutputStream bcos = null;
         try {
             if (sodaCutout.hasNoOperations()) {
                 log.debug("No parameters specified.");
                 HeadAction.setHeaders(artifact, syncOutput);
                 bcos = new ByteCountOutputStream(syncOutput.getOutputStream());
+                if (range != null) {
+                    try {
+                        SortedSet<ByteRange> rangeSet = parseRange(range, artifact.getContentLength());
+                        if (rangeSet != null) {
+                            ByteRange byteRange = rangeSet.first();
+                            syncOutput.setCode(206);
+                            long lastByte = byteRange.getOffset() + byteRange.getLength() - 1;
+                            syncOutput.setHeader(CONTENT_RANGE, "bytes " + byteRange.getOffset() + "-"
+                                    + lastByte + "/" + artifact.getContentLength());
+                            // override content length
+                            syncOutput.setHeader(CONTENT_LENGTH, byteRange.getLength());
+                            storageAdapter.get(storageLocation, bcos, rangeSet);
```

Comment:
If the code gets in here, both this storageAdapter.get(...) and the one below on line 176 are going to be executed. I think this one is buried too deeply in conditionals and instead there should be a simpler block like this:
```
if (!rangeSet.isEmpty()) {
  storageAdapter.get(storageLocation, bcos, rangeSet);
} else {
  storageAdapter.get(storageLocation, bcos);
}
```
You could hide all the conditional complexity in parseRange and always return a (possibly empty) set.


The `catch(RangeNotSatisfiableException ex)` could be in the outer (line ~220) and set appropriate headers there (you don't need another layer of try/catch in here).